File: LCP/kotlin/src/_18_早餐组合.kt
private const val Q = 1000000007

// 今天尝试了下竞赛，到第二道简单题就不行了，二分都没写出来，太菜了，😨
// 需要再练习下二分。
class _13_TowPointer {
    // 时间 O(m + n) 空间 O(1)
    fun breakfastNumber(staple: IntArray, drinks: IntArray, x: Int): Int {
        staple.sort()
        drinks.sort()
        val n = staple.size
        val m = drinks.size
        var i = 0
        var j = m - 1
        var ret = 0
        while (i < n && j >= 0) {
            if (staple[i] + drinks[j] <= x) {
                ret = (ret + j + 1) % Q
                i++
            } else {
                j--
            }
        }
        return ret
    }
}

// 唉，还是不会二分。
class _18_BinarySearch {
    // 时间 O(max(nlogn, mlogm, nlogm)) O(1)
    fun breakfastNumber(staple: IntArray, drinks: IntArray, x: Int): Int {
        staple.sort()
        drinks.sort()
        var ret = 0
        for (i in staple.indices) {
            val t = x - staple[i]
            if (t <= 0) break
            var l = 0
            var r = drinks.size
            while (l < r) {
                val m = (l + r) ushr 1
                if (drinks[m] > t) {
                    r = m
                } else {
                    l = m + 1
                }
            }
            if (l == 0) break
            ret = (ret + l) % Q
        }
        return ret
    }
}

fun main() {

}

File: LCP/cpp/src/_18_早餐组合.cpp
#include <vector>
#include <iostream>
#include <string>

using namespace std;

class UsingBinarySearch
{
public:
    int breakfastNumber(vector<int> &staple, vector<int> &drinks, int x)
    {
        sort(staple.begin(), staple.end());
        sort(drinks.begin(), drinks.end());
        const int Q = 1000000007;
        int ret = 0;
        for (int i = 0; i < staple.size(); i++)
        {
            int l = 0, r = drinks.size(), t = x - staple[i];
            while (l < r)
            {
                int m = l + ((r - l) >> 1);
                if (drinks[m] <= t)
                    l = m + 1;
                else
                    r = m;
            }
            if (l == 0)
                break;
            ret = (ret + l) % Q;
        }
        return ret;
    }
};

int main(int argc, char const *argv[])
{
    cout << "早餐组合" << endl;
    return 0;
}
